src/braket/default_simulator/density_matrix_simulation.py

```diff
@@ -0,0 +1,242 @@
+
+            if targets:
+                if isinstance(operation, (GateOperation, Observable)):
+                    dm_tensor = DensityMatrixSimulation._apply_gate(
+                        dm_tensor, qubit_count, matrix, targets
+                    )
+                if isinstance(operation, KrausOperation):
+                    dm_tensor = DensityMatrixSimulation._apply_kraus(
+                        dm_tensor, qubit_count, matrix, targets
+                    )
+            elif targets is None:
+                if isinstance(operation, (GateOperation, Observable)):
+                    for qubit in range(qubit_count):
+                        dm_tensor = DensityMatrixSimulation._apply_gate(
+                            dm_tensor, qubit_count, matrix, (qubit,)
+                        )
+                if isinstance(operation, KrausOperation):
+                    for qubit in range(qubit_count):
+                        dm_tensor = DensityMatrixSimulation._apply_kraus(
+                            dm_tensor, qubit_count, matrix, (qubit,)
+                        )
+
+        return np.reshape(dm_tensor, (2 ** qubit_count, 2 ** qubit_count))
+
+    def retrieve_samples(self) -> List[int]:
+        """ Retrieves samples of states from the density matrix of the simulation,
+        based on the probabilities.
+
+        Returns:
+            List[int]: List of states sampled according to their probabilities
+            in the density matrix. Each integer represents the decimal encoding of the
+            corresponding computational basis state.
+        """
+        return np.random.choice(
+            self._density_matrix.shape[0], p=self.probabilities, size=self._shots
+        )
+
+    @property
+    def state(self) -> np.ndarray:
+        """
+        np.ndarray: The current state of the simulation.
+        """
+        return self.density_matrix
+
+    @property
+    def density_matrix(self) -> np.ndarray:
+        """
+        np.ndarray: The density matrix specifying the current state of the simulation.
+        """
+        return self._density_matrix
+
+    @property
+    def state_with_observables(self) -> np.ndarray:
+        """
+        np.ndarray: The final density matrix of the simulation after application of observables.
+
+        Raises:
+            RuntimeError: If observables have not been applied
+        """
+        if self._post_observables is None:
+            raise RuntimeError("No observables applied")
+        return self._post_observables
+
+    @property
+    def probabilities(self) -> np.ndarray:
+        """np.ndarray: The probabilities of each computational basis state."""
+        return self.probabilities_from_state(self._density_matrix)
+
+    @staticmethod
+    def probabilities_from_state(state) -> np.ndarray:
+        """np.ndarray: The probabilities of each computational basis state of a
+        given state.
+        """
+        return np.real(np.diag(state))
```

Comment:
You can just put this into `probabilities`:

```python
@property
def probabilities(self) -> np.ndarray:
    """..."""
    return np.real(np.diag(self._density_matrix))
```

Reply:
We still need to expose a method to get probabilities from _post_observables.

I removed `probabilities_from_state` from simulation.py, state_vector_simulation.py, density_matrix_simulation.py 
and added this method as `_probabilities` in state_vector_simulation.py and density_matrix_simulation.py